src/raft_log.rs

```diff
@@ -42,20 +43,26 @@ pub struct RaftLog<T: Storage> {
     /// on a quorum of nodes.
     ///
     /// Invariant: applied <= committed
+    /// NOTE: this invariant can be break after restart if `max_apply_unpersisted_log_limit` > 0,
+    /// but once the committed catches up with applied, it should never fall behind again.
     pub committed: u64,
 
     /// The highest log position that is known to be persisted in stable
     /// storage. It's used for limiting the upper bound of committed and
     /// persisted entries.
     ///
-    /// Invariant: persisted < unstable.offset && applied <= persisted
+    /// Invariant: persisted < unstable.offset
     pub persisted: u64,
 
     /// The highest log position that the application has been instructed
     /// to apply to its state machine.
     ///
-    /// Invariant: applied <= min(committed, persisted)
+    /// Invariant: applied <= min(committed, persisted) iff `max_apply_unpersisted_log_limit` is 0.
     pub applied: u64,
+
+    /// the maximum log gap between persisted_index and applied_index.
+    /// Caller should ensure the value won't lead to the upper bound overflow.
```

Comment:
The comment about overflow may be unnecessary. Alternatively, it can be written in `Config`.
```suggestion
    /// The maximum log gap between persisted and applied.
```